x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,296 @@
+        }
+
+        private CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext, String desc, long offset, long length) {
+            super(desc, ioContext);
+            this.fileName = fileName;
+            this.fileLength = fileLength;
+            this.file = cacheDir.resolve(fileName);
+            this.ioContext = ioContext;
+            this.offset = offset;
+            this.end = offset + length;
+            this.cacheFile = new AtomicReference<>();
+            this.closed = new AtomicBoolean(false);
+
+            final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            this.cacheEvictionLock = new ReleasableLock(readWriteLock.writeLock());
+            this.cacheAcquireLock = new ReleasableLock(readWriteLock.readLock());
+        }
+
+        @Override
+        public long length() {
+            return end - offset;
+        }
+
+        @Nullable
+        private CacheFile getOrAcquire() throws Exception {
+            CacheFile currentCacheFile = cacheFile.get();
+            if (currentCacheFile != null) {
+                return currentCacheFile;
+            }
+
+            final CacheFile newCacheFile = cacheService.get(fileName, fileLength, file);
+            try (ReleasableLock ignored = cacheEvictionLock.acquire()) {
+                if (newCacheFile.acquire(this)) {
+                    final CacheFile previousCacheFile = cacheFile.getAndSet(newCacheFile);
+                    assert previousCacheFile == null;
+                    return newCacheFile;
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public void onEviction(final CacheFile evictedCacheFile) {
+            try (ReleasableLock ignored = cacheEvictionLock.acquire()) {
+                if (cacheFile.compareAndSet(evictedCacheFile, null)) {
+                    evictedCacheFile.release(this);
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            if (closed.compareAndSet(false, true)) {
+                try (ReleasableLock ignored = cacheEvictionLock.acquire()) {
+                    final CacheFile currentCacheFile = cacheFile.getAndSet(null);
+                    if (currentCacheFile != null) {
+                        currentCacheFile.release(this);
+                    }
+                }
+            }
+        }
+
+        @Override
+        protected void readInternal(final byte[] buffer, final int offset, final int length) throws IOException {
+            final long position = getFilePointer() + this.offset;
+
+            int bytesRead = 0;
+            while (bytesRead < length) {
+                final long pos = position + bytesRead;
+                final int off = offset + bytesRead;
+                final int len = length - bytesRead;
+
+                CacheFile cacheFile = null;
+                try {
+                    cacheFile = getOrAcquire();
```

Comment:
If this is a clone of an `IndexInput` over a `CacheFile` that was evicted then this will repopulate the cache and open the file channel, and then when we release the `CacheFile` later on we'll close the file channel; subsequent reads from this clone will all open and close the file channel too, which I think will be too expensive. I think we need to get the root `IndexInput` to acquire the `CacheFile` so that we keep the file channel open between reeds.

Reply:
I agree - I pushed 262d7e1cb6e123a7853682a04aab27ffb3978419 in which clones/slices now re-acquire the `CacheFile` using the root `IndexInput`. This would reduce the number of eviction listeners and should prevent too frequent opens/closes on files (unless the cache is close to be full).